src/coord/src/coord.rs

```diff
@@ -3415,51 +3733,48 @@ impl<S: Append + 'static> Coordinator<S> {
         {
             action = EndTransactionAction::Rollback;
         }
-        let response = ExecuteResponse::TransactionExited {
+        let response = Ok(ExecuteResponse::TransactionExited {
             tag: action.tag(),
             was_implicit: session.transaction().is_implicit(),
-        };
+        });
 
         // Immediately do tasks that must be serialized in the coordinator.
-        let rx = self
+        let result = self
             .sequence_end_transaction_inner(&mut session, action)
             .await;
 
-        // We can now wait for responses or errors and do any session/transaction
-        // finalization in a separate task.
-        let conn_id = session.conn_id();
-        task::spawn(
-            || format!("sequence_end_transaction:{conn_id}"),
-            async move {
-                let result = match rx {
-                    // If we have more work to do, do it
-                    Ok(fut) => fut.await,
-                    Err(e) => Err(e),
-                };
-
-                if result.is_err() {
-                    action = EndTransactionAction::Rollback;
-                }
+        match result {
+            Ok(Some(writes)) if writes.is_empty() => {
                 session.vars_mut().end_transaction(action);
-
-                match result {
-                    Ok(None) => tx.send(Ok(response), session),
-                    Ok(Some(response)) => tx.send(Ok(response), session),
-                    Err(err) => tx.send(Err(err), session),
-                }
-            },
-        );
+                tx.send(response, session);
+            }
+            Ok(Some(writes)) => self.submit_write(PendingWriteTxn::User {
+                writes: writes.into_iter().map(|write| write.into()).collect(),
+                response,
+                client_transmitter: tx,
+                session,
+                action: Some(action),
+            }),
+            Ok(None) => {
+                session.vars_mut().end_transaction(action);
```

Comment:
I am displeased that this is now called in 4 places. I think we can now have this whole match block return a `Result<ExecuteResponse, CoordError>`, then just call this and tx.send below, and then then group commit match branch does an early return to avoid that.

Reply:
Good idea, just changed it.